<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Pedigree kinship() details • Pedixplorer</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Roboto-0.4.9/font.css" rel="stylesheet">
<link href="../deps/JetBrains_Mono-0.4.9/font.css" rel="stylesheet">
<link href="../deps/Roboto_Slab-0.4.9/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="../deps/MathJax-3.2.2/tex-chtml.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Pedigree kinship() details">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Pedixplorer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/Pedixplorer.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/pedigree_alignment.html">Pedigree alignment details</a></li>
    <li><a class="dropdown-item" href="../articles/pedigree_kinship.html">Pedigree kinship() details</a></li>
    <li><a class="dropdown-item" href="../articles/pedigree_object.html">Pedigree object</a></li>
    <li><a class="dropdown-item" href="../articles/pedigree_plot.html">Pedigree plotting details</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-news" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">News</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-news">
<li><h6 class="dropdown-header" data-toc-skip>Releases</h6></li>
    <li><a class="external-link dropdown-item" href="https://github.com/LouisLeNezet/Pedixplorer/releases/tag/v1.1.5">Latest devel release : Version 1.1.5</a></li>
    <li><a class="external-link dropdown-item" href="https://www.bioconductor.org/packages/release/bioc/html/Pedixplorer.html">Latest main release : Version 1.0.0</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../news/index.html">Changelog</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/LouisLeNezet/Pedixplorer/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Pedigree kinship() details</h1>
                        <h4 data-toc-skip class="author">TM
Therneau</h4>
            
            <h4 data-toc-skip class="date">25 February, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/LouisLeNezet/Pedixplorer/blob/math/vignettes/pedigree_kinship.Rmd" class="external-link"><code>vignettes/pedigree_kinship.Rmd</code></a></small>
      <div class="d-none name"><code>pedigree_kinship.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The kinship matrix is foundational for random effects models with
family data. For <span class="math inline">\(n\)</span> subjects it is
an <span class="math inline">\(n \times n\)</span> matrix whose <span class="math inline">\(i\)</span>, <span class="math inline">\(j\)</span>
elements contains the expected fraction of alleles that would be
identical by descent if we sampled one from subject <span class="math inline">\(i\)</span> and another from subject <span class="math inline">\(j\)</span>. Note that the diagonal elements of the
matrix will be 0.5 not 1: when we randomly sample twice from the same
subject (with replacement) we will get two copies of the gene inherited
from the father 1/4 of the time, the maternal copy twice (1/4) or one of
each 1/2 the time. The formal definition is <span class="math inline">\(K(i,i) = 1/4 + 1/4 + 1/2 K(m,f)\)</span> where
<span class="math inline">\(m\)</span> and <span class="math inline">\(f\)</span> are the father and mother of subject
<span class="math inline">\(i\)</span>.</p>
<p>The algorithm used is found in K Lange, <em>Mathematical and
Statistical Methods for Genetic Analysis</em>, Springer 1997, page
71–72.</p>
<p>The key idea of the recursive algorithm for <span class="math inline">\(K(i,j)\)</span> is to condition on the gene
selection for the first index <span class="math inline">\(i\)</span>.
Let <span class="math inline">\(m(i)\)</span> and <span class="math inline">\(f(i)\)</span> be the indices of the mother and
father of subject <span class="math inline">\(i\)</span> and <span class="math inline">\(g\)</span> be the allele randomly sampled from
subject <span class="math inline">\(i\)</span>, which may of either
maternal or paternal origin.</p>
<p><span class="math display">\[
\begin{align}
    K(i,j) &amp;= P(\mbox{$g$ maternal}) * K(m(i), j) +
                P(\mbox{$g$ paternal}) * K(f(i), j) \\
            &amp;= \frac{1}{2} K(m(i), j) + \frac{1}{2} K(f(i), j) \\
    K(i,i) &amp;= \frac{1}{2}(1 + K(m(i), f(i)))
\end{align}
\]</span></p>
<p>The key step in equation <span class="math inline">\(K(i,j)\)</span>
is if <span class="math inline">\(g\)</span> has a maternal origin, then
it is a random selection from the two maternal genes, and its IBD state
with respect to subject <span class="math inline">\(j\)</span> is that
of a random selection from <span class="math inline">\(m(i)\)</span> to
a random selection from <span class="math inline">\(j\)</span>. This is
precisely the definition of <span class="math inline">\(K(m(i),
j)\)</span>. The recursion does not work for <span class="math inline">\(K(i,i)\)</span> since once we select a maternal
gene the second choice from <span class="math inline">\(j\)</span>
cannot use a different maternal gene.</p>
<p>For the recurrence algorithm to work properly we need to compute the
values of <span class="math inline">\(K\)</span> for any parent before
the calculations for their children. Pedigree founders (those with no
parents) are assumed to be unassociated, so for these subjects we
have</p>
<p><span class="math display">\[
\begin{align*}
    K(i,i) &amp;= 1/2 \\
    K(i,j) &amp;= 0 \ ; i\ne j
\end{align*}
\]</span></p>
<p>The final formula slightly different for the <span class="math inline">\(X\)</span> chromosome. Equation <span class="math inline">\(K(i,j)\)</span> still holds, but for males the
probability that a selected <span class="math inline">\(X\)</span>
chromosome is maternal is 1, so when <span class="math inline">\(i\)</span> a male the recurrence formula becomes
<span class="math inline">\(K(i,j) = K(m(i),j)\)</span>. For females it
is unchanged. All males will have <span class="math inline">\(K(i,i) =
1\)</span> for the <span class="math inline">\(X\)</span>
chromosome.</p>
<p>In order to have already-defined terms on the right hand side of the
recurrence formula for each element, subjects need to be processed in
the following order</p>
<ul>
<li>Generation 0 (founders)</li>
<li>
<span class="math inline">\(K(i,j)\)</span> where <span class="math inline">\(i\)</span> is from generation 1 and <span class="math inline">\(j\)</span> from generation 0.</li>
<li>
<span class="math inline">\(K(i,j)\)</span> with <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> from generation 1</li>
<li>
<span class="math inline">\(K(i,j)\)</span> with <span class="math inline">\(i\)</span> from generation 2 and <span class="math inline">\(j\)</span> from generation 0 or 1</li>
<li>
<span class="math inline">\(K(i,j)\)</span> with <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> from generation 2.</li>
<li>…</li>
</ul>
<p>The kindepth routine assigns a plotting depth to each subject in such
a way that parents are always above children. For each depth we need to
do the compuations of formula <span class="math inline">\(K(i,j)\)</span> twice. The first time it will get
the relationship between each subject and prior generations correct, the
second will correctly compute the values between subjects on the same
level. The computations within any stage of the above list can be
vectorized, but not those between stages.</p>
<p>Let <span class="math inline">\(indx\)</span> be the index of the
rows for the generation currently being processed, say generation <span class="math inline">\(g\)</span>. We add correct computations to the
matrix one row at a time; all of the calculations depend only on the
prior rows with the exception of the <span class="math inline">\(i,i\)</span> element. This approach leads to a for
loop containing operations on single rows/columns.</p>
<p>At one point below we use a vectorized version. It looks like the
snippet below</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">g</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">depth</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">indx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">depth</span> <span class="op">==</span> <span class="va">g</span><span class="op">)</span></span>
<span>    <span class="va">kmat</span><span class="op">[</span><span class="va">indx</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">kmat</span><span class="op">[</span><span class="va">mother</span><span class="op">[</span><span class="va">indx</span><span class="op">]</span>, <span class="op">]</span> <span class="op">+</span> <span class="va">kmat</span><span class="op">[</span><span class="va">father</span><span class="op">[</span><span class="va">indx</span><span class="op">]</span>, <span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span></span>
<span>    <span class="va">kmat</span><span class="op">[</span>, <span class="va">indx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">kmat</span><span class="op">[</span>, <span class="va">mother</span><span class="op">[</span><span class="va">indx</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="va">kmat</span><span class="op">[</span>, <span class="va">father</span><span class="op">[</span><span class="va">indx</span><span class="op">]</span>, <span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="va">indx</span><span class="op">)</span> <span class="va">kmat</span><span class="op">[</span><span class="va">j</span>, <span class="va">j</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">kmat</span><span class="op">[</span><span class="va">mother</span><span class="op">[</span><span class="va">j</span><span class="op">]</span>, <span class="va">father</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The first line computes all the values for a horizontal stripe of the
matrix. It will be correct for columns in generations <span class="math inline">\(&lt;g\)</span>, unreliable for generation <span class="math inline">\(g\)</span> with itself because of incomplete
parental relationships, and zero for higher generations. The second line
does the vertical stripe, and because of the line before it does have
the data it needs and so gets all the stripe correct. Except of course
for the diagonal elements, for which formula <span class="math inline">\(K(i,j)\)</span> does not hold. We fill those in
last. We know that vectorized calculations are always faster in R and I
was excited to figure this out. The unfortunate truth is that for this
code it hardly makes a difference, and for the X chromosome calculation
leads to impenetrable if-then-else logic.</p>
<p>The program can be called with a Pedigree, or raw data. The first
argument is <span class="math inline">\(id\)</span> instead of the more
generic <span class="math inline">\(x\)</span> for backwards
compatability with an older version of the routine. We give founders a
fake parent of subject <span class="math inline">\(n+1\)</span> who is
not related to anybody (even themself); it avoids some if-then-else
constructions.</p>
<div class="section level3">
<h3 id="with-a-pedigree-object">With a Pedigree object<a class="anchor" aria-label="anchor" href="#with-a-pedigree-object"></a>
</h3>
<p>The method for a Pedigree object is an almost trivial modification.
Since the mother and father are already indexed into the id list it has
two lines that are different, those that create mrow and drow. The other
change is that now we potentially have information available on
monozygotic twins. If there are any such, then when the second twin of a
pair is added to the matrix, we need to ensure that the pairs kinship
coefficient is set to the self-self value. This can be done after each
level is complete, but before children for that level are computed. If
there are monozygotic triples, quadruplets, etc. this computation gets
more involved.</p>
<p>The total number of monozygotic twins is always small, so it is
efficient to fix up all the monzygotic twins at each generation. A
variable <span class="math inline">\(havemz\)</span> is set to
<code>TRUE</code> if there are any, and an index array <span class="math inline">\(mzindex\)</span> is created for matrix
subscripting.</p>
</div>
<div class="section level3">
<h3 id="with-multiple-families">With multiple families<a class="anchor" aria-label="anchor" href="#with-multiple-families"></a>
</h3>
<p>For the Minnesota Family Cancer Study there are 461 families and
29114 subjects. The raw kinship matrix would be 29114 by 29114 which is
over 5 terabytes of memory, something that clearly will not work within
S. The solution is to store the overall matrix as a sparse Matrix
object. Each family forms a single block. For this study we have
<code>n &lt;- table(minnbreast$famid); sum(n*(n+1)/2)</code> or 1.07
million entries; assuming that only the lower half of each matrix is
stored. The actual size is actually smaller than this, since each family
matrix will have zeros in it — founders for instance are not related —
and those zeros are also not stored.</p>
<p>The result of each per-family call to kinship will be a symmetric
matrix. We first turn each of these into a dsCMatrix object, a sparse
symmetric form. The <span class="math inline">\(bdiag\)</span> function
is then used to paste all of these individual sparse matrices into a
single large matrix.</p>
<p>Why do we note use <code>(i in famlist)</code> below? A numeric
subscript of <code>[[9]]</code> % selects the ninth family, not the
family labeled as 9, so a numeric family id would not act as we wished.
If all of the subject ids are unique, across all families, the final
matrix is labeled with the subject id, otherwise it is labeled with
family/subject.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://louislenezet.github.io/Pedixplorer/">Pedixplorer</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">sampleped</span><span class="op">)</span></span>
<span><span class="va">pedi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">sampleped</span><span class="op">[</span>, <span class="op">-</span><span class="fl">16</span><span class="op">]</span><span class="op">)</span> <span class="co"># Remove adoption field</span></span>
<span><span class="fu"><a href="../reference/kinship.html">kinship</a></span><span class="op">(</span><span class="va">pedi</span><span class="op">)</span><span class="op">[</span><span class="fl">35</span><span class="op">:</span><span class="fl">48</span>, <span class="fl">35</span><span class="op">:</span><span class="fl">48</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 14 x 14 sparse Matrix of class "dsCMatrix"</span></span></code></pre>
<pre><code><span><span class="co">##   [[ suppressing 14 column names '1_135', '1_136', '1_137' ... ]]</span></span></code></pre>
<pre><code><span><span class="co">##                                                                                </span></span>
<span><span class="co">## 1_135 0.500 .     .    0.25 0.125 0.125 0.125 .    .    .   .    .    .    .   </span></span>
<span><span class="co">## 1_136 .     0.500 .    0.25 0.125 0.125 0.125 .    .    .   .    .    .    .   </span></span>
<span><span class="co">## 1_137 .     .     0.50 .    0.250 0.250 0.250 .    .    .   .    .    .    .   </span></span>
<span><span class="co">## 1_138 0.250 0.250 .    0.50 0.250 0.250 0.250 .    .    .   .    .    .    .   </span></span>
<span><span class="co">## 1_139 0.125 0.125 0.25 0.25 0.500 0.250 0.250 .    .    .   .    .    .    .   </span></span>
<span><span class="co">## 1_140 0.125 0.125 0.25 0.25 0.250 0.500 0.250 .    .    .   .    .    .    .   </span></span>
<span><span class="co">## 1_141 0.125 0.125 0.25 0.25 0.250 0.250 0.500 .    .    .   .    .    .    .   </span></span>
<span><span class="co">## 2_201 .     .     .    .    .     .     .     0.50 .    .   0.25 0.25 0.25 0.25</span></span>
<span><span class="co">## 2_202 .     .     .    .    .     .     .     .    0.50 .   0.25 0.25 0.25 0.25</span></span>
<span><span class="co">## 2_203 .     .     .    .    .     .     .     .    .    0.5 .    .    .    .   </span></span>
<span><span class="co">## 2_204 .     .     .    .    .     .     .     0.25 0.25 .   0.50 0.25 0.25 0.25</span></span>
<span><span class="co">## 2_205 .     .     .    .    .     .     .     0.25 0.25 .   0.25 0.50 0.25 0.25</span></span>
<span><span class="co">## 2_206 .     .     .    .    .     .     .     0.25 0.25 .   0.25 0.25 0.50 0.25</span></span>
<span><span class="co">## 2_207 .     .     .    .    .     .     .     0.25 0.25 .   0.25 0.25 0.25 0.50</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="makekinship">MakeKinship<a class="anchor" aria-label="anchor" href="#makekinship"></a>
</h3>
<p>The older <span class="math inline">\(makekinship\)</span> function,
from before the creation of PedigreeList objects, accepts the raw
identifier data, along with a special family code for unrelated
subjects, as produced by the <code><a href="../reference/make_famid.html">make_famid()</a></code> function. All the
unrelated subjects are put at the front of the kinship matrix in this
case rather than within the family. Because unrelateds get put into a
fake family, we cannot create a rational family/subject identifier; the
id must be unique across families.</p>
<p>We include a copy of the routine for backwards compatability, but do
not anticipate any new usage of it. Like most routines, this starts out
with a collection of error checks.</p>
</div>
<div class="section level3">
<h3 id="monozygotic-twins-for-pedigree-object">Monozygotic twins, for Pedigree object<a class="anchor" aria-label="anchor" href="#monozygotic-twins-for-pedigree-object"></a>
</h3>
<p>Return now to the question of monzygotic sets, used specifically in
the kinship for Pedigree objects. Consider the following rather
difficult example:</p>
<table class="table">
<thead><tr class="header">
<th>id1</th>
<th>id2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>2</td>
</tr>
<tr class="even">
<td>2</td>
<td>3</td>
</tr>
<tr class="odd">
<td>5</td>
<td>6</td>
</tr>
<tr class="even">
<td>3</td>
<td>7</td>
</tr>
<tr class="odd">
<td>10</td>
<td>9</td>
</tr>
</tbody>
</table>
<p>Subjects 1, 2, 3, and 7 form a monozygotic quadruple, 5/6 and 9/10
are monzygotic pairs. First create a vector <code>mzgrp</code> which
contains for each subject the lowest index of a monozygotic twin for
that subject. For non-twins it can have any value. For this example that
vector is set to 1 for subjects 1, 2, 3, and 7, to 5 for 5 and 6, and to
9 for 9 and 10. Creating this requires a short while loop. Once this is
in hand we can identify the sets.</p>
<p>Next make a matrix that has a row for every possible pair. Finally,
remove the rows that are identical. The result is a set of all pairs of
observations in the matrix that correspond to monozygotic pairs.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">7</span>, <span class="fl">8</span><span class="op">)</span>,</span>
<span>    dadid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="cn">NA</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="cn">NA</span><span class="op">)</span>,</span>
<span>    momid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">8</span>, <span class="fl">8</span>, <span class="cn">NA</span>, <span class="fl">8</span>, <span class="fl">8</span>, <span class="fl">8</span>, <span class="cn">NA</span><span class="op">)</span>,</span>
<span>    sex <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">rel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    id1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">6</span>, <span class="fl">7</span><span class="op">)</span>,</span>
<span>    id2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>    code <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pedi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">rel</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pedi</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_kinship_files/figure-html/kinship_twins-1.png" alt="Monozygotic twins" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">twins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">7</span>, <span class="fl">5</span>, <span class="fl">6</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/kinship.html">kinship</a></span><span class="op">(</span><span class="va">pedi</span><span class="op">)</span><span class="op">[</span><span class="va">twins</span>, <span class="va">twins</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 6 x 6 sparse Matrix of class "dsCMatrix"</span></span>
<span><span class="co">##      1    2    3    7    5    6</span></span>
<span><span class="co">## 1 0.50 0.50 0.50 0.50 0.25 0.25</span></span>
<span><span class="co">## 2 0.50 0.50 0.50 0.50 0.25 0.25</span></span>
<span><span class="co">## 3 0.50 0.50 0.50 0.50 0.25 0.25</span></span>
<span><span class="co">## 7 0.50 0.50 0.50 0.50 0.25 0.25</span></span>
<span><span class="co">## 5 0.25 0.25 0.25 0.25 0.50 0.50</span></span>
<span><span class="co">## 6 0.25 0.25 0.25 0.25 0.50 0.50</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] Pedixplorer_1.3.2 BiocStyle_2.32.1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] gtable_0.3.6          xfun_0.51             bslib_0.9.0          </span></span>
<span><span class="co">##  [4] ggplot2_3.5.1         htmlwidgets_1.6.4     lattice_0.22-6       </span></span>
<span><span class="co">##  [7] quadprog_1.5-8        vctrs_0.6.5           tools_4.4.2          </span></span>
<span><span class="co">## [10] generics_0.1.3        stats4_4.4.2          tibble_3.2.1         </span></span>
<span><span class="co">## [13] pkgconfig_2.0.3       Matrix_1.7-2          data.table_1.17.0    </span></span>
<span><span class="co">## [16] desc_1.4.3            S4Vectors_0.42.1      readxl_1.4.3         </span></span>
<span><span class="co">## [19] lifecycle_1.0.4       compiler_4.4.2        stringr_1.5.1        </span></span>
<span><span class="co">## [22] shinytoastr_2.2.0     textshaping_1.0.0     munsell_0.5.1        </span></span>
<span><span class="co">## [25] httpuv_1.6.15         shinyWidgets_0.9.0    htmltools_0.5.8.1    </span></span>
<span><span class="co">## [28] sass_0.4.9            yaml_2.3.10           lazyeval_0.2.2       </span></span>
<span><span class="co">## [31] plotly_4.10.4         later_1.4.1           pillar_1.10.1        </span></span>
<span><span class="co">## [34] pkgdown_2.1.1         jquerylib_0.1.4       tidyr_1.3.1          </span></span>
<span><span class="co">## [37] DT_0.33               cachem_1.1.0          mime_0.12            </span></span>
<span><span class="co">## [40] tidyselect_1.2.1      digest_0.6.37         stringi_1.8.4        </span></span>
<span><span class="co">## [43] colourpicker_1.3.0    dplyr_1.1.4           purrr_1.0.4          </span></span>
<span><span class="co">## [46] bookdown_0.42         fastmap_1.2.0         grid_4.4.2           </span></span>
<span><span class="co">## [49] colorspace_2.1-1      cli_3.6.4             magrittr_2.0.3       </span></span>
<span><span class="co">## [52] withr_3.0.2           scales_1.3.0          promises_1.3.2       </span></span>
<span><span class="co">## [55] rmarkdown_2.29        httr_1.4.7            gridExtra_2.3        </span></span>
<span><span class="co">## [58] cellranger_1.1.0      ragg_1.3.3            shiny_1.10.0         </span></span>
<span><span class="co">## [61] evaluate_1.0.3        knitr_1.49            shinycssloaders_1.1.0</span></span>
<span><span class="co">## [64] miniUI_0.1.1.1        viridisLite_0.4.2     rlang_1.1.5          </span></span>
<span><span class="co">## [67] Rcpp_1.0.14           xtable_1.8-4          glue_1.8.0           </span></span>
<span><span class="co">## [70] BiocManager_1.30.25   BiocGenerics_0.50.0   jsonlite_1.9.0       </span></span>
<span><span class="co">## [73] R6_2.6.1              plyr_1.8.9            systemfonts_1.2.1    </span></span>
<span><span class="co">## [76] fs_1.6.5</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Louis Le Nezet, Jason Sinnwell, Terry Therneau.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
